test/discrete/test_il_bcq.py

```diff
@@ -93,10 +94,30 @@ def save_fn(policy):
     def stop_fn(mean_rewards):
         return mean_rewards >= env.spec.reward_threshold
 
+    def save_train_fn(epoch, env_step, gradient_step):
+        # see also: https://pytorch.org/tutorials/beginner/saving_loading_models.html
+        torch.save({
+            'model': policy.state_dict(),
+            'optim': optim.state_dict(),
+        }, os.path.join(log_path, 'checkpoint.pth'))
+
+    if args.resume:
+        # load from existing checkpoint
+        print(f"Loading agent under {log_path}")
+        ckpt_path = os.path.join(log_path, 'checkpoint.pth')
+        if os.path.exists(ckpt_path):
+            checkpoint = torch.load(ckpt_path, map_location=args.device)
+            policy.load_state_dict(checkpoint['model'])
+            # optim.load_state_dict(checkpoint['optim'])  # don't know why
```

Comment:
This is strange. If I load optimizer from an existing checkpoint, it doesn't work:
```python
$ python3 test/discrete/test_il_bcq.py --resume                                                                                                                                                          [15:03:52]
2021-04-25 15:05:14.273532: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
Loading agent under log/CartPole-v0/discrete_bcq
Successfully restore policy and optim.
Epoch #2:   0%|                                                                                                                                                                            | 0/2000 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "test/discrete/test_il_bcq.py", line 142, in <module>
    test_discrete_bcq(get_args())
  File "test/discrete/test_il_bcq.py", line 116, in test_discrete_bcq
    result = offline_trainer(
  File "/home/trinkle/github/tianshou-new/tianshou/trainer/offline.py", line 94, in offline_trainer
    losses = policy.update(batch_size, buffer)
  File "/home/trinkle/github/tianshou-new/tianshou/policy/base.py", line 235, in update
    result = self.learn(batch, **kwargs)
  File "/home/trinkle/github/tianshou-new/tianshou/policy/imitation/discrete_bcq.py", line 117, in learn
    self.optim.step()
  File "/home/trinkle/.local/lib/python3.8/site-packages/torch/optim/optimizer.py", line 89, in wrapper
    return func(*args, **kwargs)
  File "/home/trinkle/.local/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/trinkle/.local/lib/python3.8/site-packages/torch/optim/adam.py", line 108, in step
    F.adam(params_with_grad,
  File "/home/trinkle/.local/lib/python3.8/site-packages/torch/optim/_functional.py", line 96, in adam
    param.addcdiv_(exp_avg, denom, value=-step_size)
RuntimeError: output with shape [64] doesn't match the broadcast shape [64, 64]
```

Reply:
This is because adam use `set(actor.parameters()).union(critic_parameters())` which causes random ordering in adam param groups.